File: demo/src/lib/server/verify.ts
import "server-only";
import { decodeFunctionData, parseEventLogs, type Address, type Hex } from "viem";
import { cctpAdapterAbi, oftAdapterAbi } from "@/lib/abis/generated";
import { isSupportedChain, type SupportedChainId } from "@/lib/chains";
import { addressToBytes32, BRIDGE_CCTP, orderIdOf, type Order } from "@/lib/order";
import { getToken } from "@/lib/tokens";
import { pub } from "./clients";
import { adapterAddressServer } from "./env";
import { assertOrderAllowed } from "./guards";

export interface VerifiedOrder {
  order: Order;
  orderId: Hex;
  bridgeType: number;
  srcTxHash: Hex;
  srcBlockNumber: bigint;
}

function configuredAdapters(): { addr: Address; bridgeType: number }[] {
  const out: { addr: Address; bridgeType: number }[] = [];
  for (const bt of [0, 1]) {
    try {
      out.push({ addr: adapterAddressServer(bt), bridgeType: bt });
    } catch {
      /* adapter not configured (e.g. no OFT) */
    }
  }
  return out;
}

/** Pull the initiate call out of the tx input, unwrapping a `multicall([selfPermit, initiate])`. */
function findInitiate(input: Hex, abi: typeof cctpAdapterAbi | typeof oftAdapterAbi) {
  const decoded = decodeFunctionData({ abi, data: input });
  if (decoded.functionName === "multicall") {
    const calls = (decoded.args as readonly [readonly Hex[]])[0];
    for (const c of calls) {
      try {
        const inner = decodeFunctionData({ abi, data: c });
        if (inner.functionName.startsWith("initiate")) return inner;
      } catch {
        /* not an adapter call */
      }
    }
    throw new Error("no initiate call found inside multicall");
  }
  if (decoded.functionName.startsWith("initiate")) return decoded;
  throw new Error(`unexpected source function: ${decoded.functionName}`);
}

/**
 * Reconstruct the full Order from a source initiate tx and PROVE its authenticity: the recomputed
 * `orderId` must equal the `orderId` emitted by OUR adapter's `OrderCreated` in that tx. A relayer
 * that fills any order failing this check is donating funds — so this is the one mandatory gate.
 */
export async function reconstructAndVerify(srcChainId: number, txHash: Hex): Promise<VerifiedOrder> {
  if (!isSupportedChain(srcChainId)) throw new Error(`unsupported source chain ${srcChainId}`);
  const src = srcChainId as SupportedChainId;
  const client = pub(src);

  const [receipt, tx] = await Promise.all([
    client.getTransactionReceipt({ hash: txHash }),
    client.getTransaction({ hash: txHash }),
  ]);
  if (receipt.status !== "success") throw new Error("source transaction reverted");

  // Find an OrderCreated emitted by one of our adapters (pins the source contract + bridge family).
  const adapters = configuredAdapters();
  const events = parseEventLogs({ abi: cctpAdapterAbi, eventName: "OrderCreated", logs: receipt.logs });
  const match = events
    .map((e) => ({ e, hit: adapters.find((a) => a.addr.toLowerCase() === (e.address as string).toLowerCase()) }))
    .find((x) => x.hit);
  if (!match || !match.hit) throw new Error("no OrderCreated from a known fast-fill adapter in this tx");

  const bridgeType = match.hit.bridgeType;
  const ev = match.e.args as {
    orderId: Hex;
    bridgeType: number;
    sender: Address;
    dstChainId: number;
    outputToken: Hex;
    outputAmount: bigint;
    nonce: bigint;
  };
  if (Number(ev.bridgeType) !== bridgeType) throw new Error("bridgeType mismatch (adapter vs event)");

  // Decode the initiate args. CCTP has an extra `mintFee` after `maxFee`; OFT keeps its original shape.
  const abi = bridgeType === BRIDGE_CCTP ? cctpAdapterAbi : oftAdapterAbi;
  const init = findInitiate(tx.input, abi);
  const a = init.args as readonly unknown[];
  const dstChainId = Number(a[0] as number | bigint);
  if (!isSupportedChain(dstChainId)) throw new Error(`unsupported destination chain ${dstChainId}`);
  const dst = dstChainId as SupportedChainId;

  const recipient = a[1] as Hex; // bytes32
  const inputAmount = a[2] as bigint;
  const outputAmount =
    bridgeType === BRIDGE_CCTP ? inputAmount - (a[3] as bigint) - (a[4] as bigint) : (a[3] as bigint);
  const deliveryWindow = (bridgeType === BRIDGE_CCTP ? a[6] : a[5]) as bigint;
  const discountRate = (bridgeType === BRIDGE_CCTP ? a[7] : a[6]) as bigint;
  const baseFee = (bridgeType === BRIDGE_CCTP ? a[8] : a[7]) as bigint;
  const exec = (bridgeType === BRIDGE_CCTP ? a[9] : a[8]) as { gasLimit: bigint; data: Hex }; // Execution

  const symbol = bridgeType === BRIDGE_CCTP ? "USDC" : "USDT0";
  const block = await client.getBlock({ blockNumber: receipt.blockNumber });

  const order: Order = {
    bridgeType,
    srcChainId: src,
    dstChainId: dst,
    sender: addressToBytes32(ev.sender),
    recipient,
    inputToken: addressToBytes32(getToken(src, symbol).address),
    outputToken: addressToBytes32(getToken(dst, symbol).address),
    inputAmount,
    outputAmount,
    nonce: ev.nonce,
    startTime: block.timestamp,
    expectedDeliveryTime: block.timestamp + deliveryWindow,
    discountRate,
    baseFee,
    callbackGasLimit: exec.gasLimit,
    hookData: exec.data,
  };

  const orderId = orderIdOf(order);
  if (orderId.toLowerCase() !== ev.orderId.toLowerCase()) {
    throw new Error("authenticity check failed: recomputed orderId does not match OrderCreated");
  }

  assertOrderAllowed(order);
  return { order, orderId, bridgeType, srcTxHash: txHash, srcBlockNumber: receipt.blockNumber };
}

File: relayer/bot/src/verify.rs
//! Reconstruct the full `Order` from a source `initiate*` tx and prove its authenticity.
//!
//! Port of `demo/src/lib/server/verify.ts`. The `OrderCreated` event carries only 7 of 16 fields;
//! the rest come from the tx calldata (unwrapping `multicall([selfPermit, initiate])`), the source
//! block timestamp (`startTime`), and the token registry. The recomputed `orderId` MUST equal the
//! emitted one — fill against a wrong order and the bot is never reimbursed.

use alloy::consensus::Transaction as _;
use alloy::primitives::{Address, Bytes, FixedBytes, B256, U256};
use alloy::providers::Provider;
use alloy::rpc::types::BlockNumberOrTag;
use alloy::sol_types::SolCall;
use eyre::{eyre, Result, WrapErr};

use crate::app::App;
use crate::config::BRIDGE_CCTP;
use crate::order::{addr_to_b32, order_id};
use crate::sol::{cctp, oft, Order};
use crate::watcher::Discovered;

#[derive(Clone)]
#[allow(dead_code)] // src_block_number retained for provenance
pub struct VerifiedOrder {
    pub order: Order,
    pub order_id: B256,
    pub bridge_type: u8,
    pub oft_id: Option<u8>,
    pub adapter: Address,
    pub src_chain_id: u64,
    pub dst_chain_id: u64,
    pub src_tx_hash: B256,
    pub src_block_number: u64,
    /// CCTP `mintFee` (USDC base units). `> 0` ⇒ executor-routed (settle via `CctpExecutor.execute`,
    /// the mint-relay path that pays this fee). `0` ⇒ direct (settle via `CctpAdapter.settle`). OFT: 0.
    pub mint_fee: U256,
}

/// The leading `initiate*` args we need (positions are identical for the `*For` variants).
struct InitiateArgs {
    dst_chain_id: u32,
    recipient: FixedBytes<32>,
    input_amount: U256,
    fourth: U256,   // maxFee (CCTP) or minAmountLD (OFT)
    mint_fee: U256, // CCTP executor mint-relay fee (0 for OFT / direct CCTP)
    delivery_window: u64,
    discount_rate: U256,
    base_fee: U256,
    gas_limit: u64,
    hook_data: Bytes,
}

pub async fn reconstruct_and_verify(app: &App, d: &Discovered) -> Result<VerifiedOrder> {
    let meta = app
        .registry
        .adapter_meta
        .get(&d.adapter)
        .ok_or_else(|| eyre!("OrderCreated from unknown adapter {}", d.adapter))?;
    let bridge_type = meta.bridge_type;
    let oft_id = meta.oft_id;
    if bridge_type != d.event.bridge_type {
        return Err(eyre!(
            "bridgeType mismatch: adapter={} event={}",
            bridge_type,
            d.event.bridge_type
        ));
    }

    let provider = app.chains.provider(d.src_chain_id)?;
    let tx = provider
        .get_transaction_by_hash(d.tx_hash)
        .await?
        .ok_or_else(|| eyre!("source tx {} not found", d.tx_hash))?;
    let block = provider
        .get_block_by_number(BlockNumberOrTag::Number(d.block_number))
        .await?
        .ok_or_else(|| eyre!("source block {} not found", d.block_number))?;
    let start_time = block.header.timestamp;

    let input = tx.input().clone();
    let args = decode_initiate(bridge_type, input.as_ref())?;
    let dst_chain_id = args.dst_chain_id as u64;

    let in_info = app
        .registry
        .output_token(d.src_chain_id, bridge_type, oft_id)
        .ok_or_else(|| eyre!("no input token for src chain {}", d.src_chain_id))?;
    let out_info = app
        .registry
        .output_token(dst_chain_id, bridge_type, oft_id)
        .ok_or_else(|| eyre!("no output token for dst chain {}", dst_chain_id))?;

    // CCTP reserves both the bridge fee (maxFee) and the executor mint-relay fee (mintFee); OFT uses
    // the signed slippage floor (minAmountLD).
    let output_amount = if bridge_type == BRIDGE_CCTP {
        args.input_amount - args.fourth - args.mint_fee
    } else {
        args.fourth
    };

    let order = Order {
        bridgeType: bridge_type,
        srcChainId: d.src_chain_id as u32,
        dstChainId: args.dst_chain_id,
        sender: addr_to_b32(d.event.sender),
        recipient: args.recipient,
        inputToken: addr_to_b32(in_info.token),
        outputToken: addr_to_b32(out_info.token),
        inputAmount: args.input_amount,
        outputAmount: output_amount,
        nonce: d.event.nonce,
        startTime: start_time,
        expectedDeliveryTime: start_time + args.delivery_window,
        discountRate: args.discount_rate,
        baseFee: args.base_fee,
        callbackGasLimit: args.gas_limit,
        hookData: args.hook_data,
    };

    let id = order_id(&order);
    if id != d.event.order_id {
        return Err(eyre!(
            "authenticity check failed: recomputed {} != emitted {}",
            id,
            d.event.order_id
        ));
    }
    if addr_to_b32(out_info.token) != d.event.output_token {
        return Err(eyre!("output token mismatch vs OrderCreated event"));
    }

    Ok(VerifiedOrder {
        order,
        order_id: id,
        bridge_type,
        oft_id,
        adapter: d.adapter,
        src_chain_id: d.src_chain_id,
        dst_chain_id,
        src_tx_hash: d.tx_hash,
        src_block_number: d.block_number,
        mint_fee: args.mint_fee,
    })
}

/// Decode the `initiate*` call from tx input, unwrapping a `multicall([selfPermit, initiate])`.
fn decode_initiate(bridge_type: u8, input: &[u8]) -> Result<InitiateArgs> {
    if input.len() >= 4 {
        let sel: [u8; 4] = input[0..4].try_into().unwrap();
        if sel == cctp::multicallCall::SELECTOR {
            let mc = cctp::multicallCall::abi_decode(input).wrap_err("decode multicall")?;
            for call in &mc.data {
                if let Ok(a) = decode_initiate_inner(bridge_type, call.as_ref()) {
                    return Ok(a);
                }
            }
            return Err(eyre!("no initiate call found inside multicall"));
        }
    }
    decode_initiate_inner(bridge_type, input)
}

fn decode_initiate_inner(bridge_type: u8, data: &[u8]) -> Result<InitiateArgs> {
    if data.len() < 4 {
        return Err(eyre!("calldata too short"));
    }
    let sel: [u8; 4] = data[0..4].try_into().unwrap();

    if bridge_type == BRIDGE_CCTP {
        if sel == cctp::initiateCCTPCall::SELECTOR {
            let c = cctp::initiateCCTPCall::abi_decode(data)?;
            Ok(InitiateArgs {
                dst_chain_id: c.dstChainId,
                recipient: c.recipient,
                input_amount: c.inputAmount,
                fourth: c.maxFee,
                mint_fee: c.mintFee,
                delivery_window: c.deliveryWindow,
                discount_rate: c.discountRate,
                base_fee: c.baseFee,
                gas_limit: c.exec.gasLimit,
                hook_data: c.exec.data,
            })
        } else if sel == cctp::initiateCCTPForCall::SELECTOR {
            let c = cctp::initiateCCTPForCall::abi_decode(data)?;
            Ok(InitiateArgs {
                dst_chain_id: c.dstChainId,
                recipient: c.recipient,
                input_amount: c.inputAmount,
                fourth: c.maxFee,
                mint_fee: c.mintFee,
                delivery_window: c.deliveryWindow,
                discount_rate: c.discountRate,
                base_fee: c.baseFee,
                gas_limit: c.exec.gasLimit,
                hook_data: c.exec.data,
            })
        } else {
            Err(eyre!("calldata is not a CCTP initiate call"))
        }
    } else if sel == oft::initiateOFTCall::SELECTOR {
        let c = oft::initiateOFTCall::abi_decode(data)?;
        Ok(InitiateArgs {
            dst_chain_id: c.dstChainId,
            recipient: c.recipient,
            input_amount: c.inputAmount,
            fourth: c.minAmountLD,
            mint_fee: U256::ZERO,
            delivery_window: c.deliveryWindow,
            discount_rate: c.discountRate,
            base_fee: c.baseFee,
            gas_limit: c.exec.gasLimit,
            hook_data: c.exec.data,
        })
    } else if sel == oft::initiateOFTForCall::SELECTOR {
        let c = oft::initiateOFTForCall::abi_decode(data)?;
        Ok(InitiateArgs {
            dst_chain_id: c.dstChainId,
            recipient: c.recipient,
            input_amount: c.inputAmount,
            fourth: c.minAmountLD,
            mint_fee: U256::ZERO,
            delivery_window: c.deliveryWindow,
            discount_rate: c.discountRate,
            base_fee: c.baseFee,
            gas_limit: c.exec.gasLimit,
            hook_data: c.exec.data,
        })
    } else {
        Err(eyre!("calldata is not an OFT initiate call"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::{BRIDGE_CCTP, CCTP_ADAPTER, FASTFILL_CONFIG};
    use crate::order::order_id;
    use crate::sol::cfg;
    use alloy::providers::ProviderBuilder;
    use alloy::sol_types::SolEvent;

    /// Reconstruct a REAL mainnet CCTP order from its source tx and assert the recomputed `orderId`
    /// equals the emitted one — the load-bearing authenticity gate, including the `mintFee` term in
    /// `outputAmount = inputAmount - maxFee - mintFee`.
    async fn check_golden(rpc: &str, src_chain_id: u64, tx_hash_str: &str) {
        let provider = ProviderBuilder::new()
            .connect_http(rpc.parse().unwrap())
            .erased();
        let tx_hash: B256 = tx_hash_str.parse().unwrap();

        let receipt = provider
            .get_transaction_receipt(tx_hash)
            .await
            .unwrap()
            .expect("receipt");
        let tx = provider
            .get_transaction_by_hash(tx_hash)
            .await
            .unwrap()
            .expect("tx");
        let block = provider
            .get_block_by_number(BlockNumberOrTag::Number(receipt.block_number.unwrap()))
            .await
            .unwrap()
            .unwrap();
        let start_time = block.header.timestamp;

        let ev = receipt
            .inner
            .logs()
            .iter()
            .find_map(|l| {
                if l.inner.address != CCTP_ADAPTER {
                    return None;
                }
                cctp::OrderCreated::decode_log(&l.inner).ok()
            })
            .expect("OrderCreated from CctpAdapter");
        let e = ev.data;

        let args = decode_initiate(BRIDGE_CCTP, tx.input().as_ref()).unwrap();
        let dst = args.dst_chain_id as u64;

        // FastFillConfig is a registry of ALL chains, readable from any chain's deployment.
        let conf = cfg::new(FASTFILL_CONFIG, provider.clone());
        let src_usdc = conf
            .chainConfig(U256::from(src_chain_id))
            .call()
            .await
            .unwrap()
            .usdc;
        let dst_usdc = conf.chainConfig(U256::from(dst)).call().await.unwrap().usdc;

        let order = Order {
            bridgeType: BRIDGE_CCTP,
            srcChainId: src_chain_id as u32,
            dstChainId: args.dst_chain_id,
            sender: addr_to_b32(e.sender),
            recipient: args.recipient,
            inputToken: addr_to_b32(src_usdc),
            outputToken: addr_to_b32(dst_usdc),
            inputAmount: args.input_amount,
            outputAmount: args.input_amount - args.fourth - args.mint_fee,
            nonce: e.nonce,
            startTime: start_time,
            expectedDeliveryTime: start_time + args.delivery_window,
            discountRate: args.discount_rate,
            baseFee: args.base_fee,
            callbackGasLimit: args.gas_limit,
            hookData: args.hook_data,
        };

        let id = order_id(&order);
        assert_eq!(
            id, e.orderId,
            "reconstructed orderId must equal the emitted OrderCreated.orderId"
        );
    }

    /// Direct CCTP order (`mintFee = 0`), Base → Arbitrum. Source: DEPLOYMENTS.md smoke.
    /// Network test — run with `cargo test -- --ignored` (set RPC_URL_8453 to override).
    #[tokio::test]
    #[ignore = "network: requires a Base RPC; run with --ignored"]
    async fn golden_orderid_base_cctp_direct() {
        let rpc =
            std::env::var("RPC_URL_8453").unwrap_or_else(|_| "https://mainnet.base.org".into());
        check_golden(
            &rpc,
            8453,
            "0x4c6b282cde6fc03bcc0c3bc75b40a601aaadf232e789db0643039052a7b790bc",
        )
        .await;
    }

    /// Executor-routed CCTP order (`mintFee > 0`), Arbitrum → Base. Source: DEPLOYMENTS.md smoke.
    /// Exercises the `outputAmount = inputAmount - maxFee - mintFee` reconstruction.
    #[tokio::test]
    #[ignore = "network: requires an Arbitrum RPC; run with --ignored"]
    async fn golden_orderid_arbitrum_cctp_routed() {
        let rpc = std::env::var("RPC_URL_42161")
            .unwrap_or_else(|_| "https://arbitrum-one.publicnode.com".into());
        check_golden(
            &rpc,
            42161,
            "0x83eef9a71fae6f0057ae954cc9c22221866c469f97deef8c9b398c733aa98504",
        )
        .await;
    }
}
